### PR TITLE
Fix nullable array issue 31674 in AvroGenericRecordToStorageApiProto

### LINKED ISSUE #31674: [Bug]: Issue with AvroGenericRecordToStorageApiProto.java handling nullable arrays
### What happened?

I've encountered an issue with the `AvroGenericRecordToStorageApiProto.java` class in Apache Beam, which is responsible for converting an Avro generic record into a proto object for writing to BigQuery using the Storage API.

The class contains two primary methods for handling the conversion:

1. `private TableFieldSchema fieldDescriptorFromAvroField(Schema.Field field)`
2. `private static Object toProtoValue(FieldDescriptor fieldDescriptor, Schema avroSchema, Object value)`

These methods handle array types effectively when the array in the Avro schema is defined as follows:

```json
{
  "name": "simple_array",
  "type": "array",
  "items": "string",
  "default": []
}
```

However, the application crashes when a nullable array is used, such as:

```json
{
  "name": "simple_array",
  "type": [
    "null",
    {
      "type": "array",
      "items": "string"
    }
  ],
  "default": null
}
```

This issue seems to stem from the handling of nullable arrays in the conversion methods. 

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by codertimu**
fixes #31674 

There is an issue with the `AvroGenericRecordToStorageApiProto.java` class in Apache Beam, which is responsible for converting an Avro generic record into a proto object for writing to BigQuery using the Storage API.

The funtion `private TableFieldSchema fieldDescriptorFromAvroField(Schema.Field field)` handles the array types effectively when the array in the Avro schema is defined as follows:

```json
{
  "name": "anullablearray",
  "type": "array",
  "items": "string",
  "default": []
}
```

it produces such a proto field as expected:

```proto
name: "De0c5bb51_aded_475d_946d_6ece875416a8"
field {
  name: "anullablearray"
  number: 1
  label: LABEL_REPEATED
  type: TYPE_STRING
}
```

However, it doesn't preserve the type `array` and uses the element type of the array when a nullable array is used, such as:

```json
{
  "name": "anullablearray",
  "type": [
    "null",
    {
      "type": "array",
      "items": "string"
    }
  ],
  "default": null
}
```

in this case, it produces such a proto field which is not expected and wrong:

```proto
name: "D94a1b8de_41fa_4d59_b308_999b9fc12f0a"
field {
  name: "anullablearray"
  number: 1
  label: LABEL_OPTIONAL
  type: TYPE_STRING
}
```

While handling the union, the function overwrites the array element type with nullability and results in this behaviour. I believe, it can be easily fixed by setting the `mode` with the union field schema in the `TableSchemaBuilder` while constructing the schema.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by codertimu** [comment]
hey @ahmedabu98, 
I've run the failed IT test in PreCommit Java GCP IO Direct / beam_PreCommit_Java_GCP_IO_Direct on my own gcp project and it successfully passed but in the pipeline it throws a timeout exception: ```
org.apache.beam.sdk.io.gcp.bigquery.FileLoadsStreamingIT > testDynamicDestinationsWithAutoShardingAndCopyJobs[0] FAILED
    java.net.SocketTimeoutException at FileLoadsStreamingIT.java:300
```
Could it be a transient error? I don't see a way to just rerun the failing step. Should I just do a whitespace commit to trigger the pipeline again?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java [hunk 1]
```diff
@@ -567,4 +579,30 @@ public void testMessageFromGenericRecordWithMap() throws Exception {
     }
     assertEquals(mapData, actualMap);
   }
+
+  @Test
+  public void testMessageFromGenericRecordWithNullableArray() throws Exception {
+    ImmutableList<String> aList = ImmutableList.of("one", "two", "red", "blue");
+    GenericRecord recordWithMap =
+        new GenericRecordBuilder(SCHEMA_WITH_NULLABLE_ARRAY).set("aNullableArray", aList).build();
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:587]: For completion, can we also include in this test a record with null value for the array?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java [hunk 2]
```diff
@@ -269,6 +269,7 @@ enum TestEnum {
           .endRecord();
 
   private static final Schema SCHEMA_WITH_MAP;
+  private static final Schema SCHEMA_WITH_NULLABLE_ARRAY;
```
Comment (ahmedabu98) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:272]: Out of curiosity, is there a benefit to declaring these variables here but initializing them later in a `static { }` block? Asking because we conventionally initialize them in the same line.

**Comment by ahmedabu98** [comment]
Hey @codertimu! That test is unrelated to the changes you're making here, so my guess is it's a flake. I triggered another run from my side, so we'll see if it's persistent

**Review by ahmedabu98 — COMMENTED** [review]
Thanks for making another fix here! left a couple comments

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java (hunk 1 above)
Comment (codertimu) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:587]: sure.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java (hunk 2 above)
Comment (codertimu) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java:272]: I don't think so, it must be something with refactoring I did. correcting them.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java (hunk 1 above)
Comment (codertimu) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProtoTest.java]: It was good suggestion to add this test case, helped me catch an error case. Added a fix for it.

**Review by ahmedabu98 — APPROVED** [review]
Looks great, thank you!
